### DISCUSSION TITLE
[Feature request] A node API to re-render the page from page.content

### DISCUSSION THREAD

**Discussion by DreamOfIce** [vuepress/core#1309] [Ideas]
### Clear and concise description of the problem

如题，正在开发的插件需要在`extendsPageOptions`或者`extendsPage`阶段注入一个组件，但此时页面已渲染完成，目前尝试可行的只有如下方法才能成功修改内容：
```ts
import { renderPageContent } from "@vuepress/core";

const { content, filePath, filePathRelative, frontmatter, path } = page;
Object.assign(page, renderPageContent({
    app,
    content,
    filePath,
    filePathRelative,
    options: {
          path,
          frontmatter,
          content,
          ...(filePath ? { filePath } : {}),
    },
}));
```

顺带一提，不知为何这样会破坏 vuepress-plugin-md-enhance 的 import 功能

### Suggested solution

添加对应的node API

**Comment by Mister-Hope** [comment]
> 如题，正在开发的插件需要在extendsPageOptions或者extendsPage阶段注入一个组件，但此时页面已渲染完成，目前尝试可行的只有如下方法才能成功修改内容：

修改 @mdit-vue/plugin-sfc 的结果的确是一个好的解决办法。

> 顺带一提，不知为何这样会破坏 vuepress-plugin-md-enhance 的 import 功能。

你可以提一个 reproduction，如果你说的是 include 功能，其依赖于 `@md-it/plugin-include`，你可以尝试做一个 reproduction 深入研究一下。至少我认为我写的是一个标准的 markdown-it 插件。

**Comment by DreamOfIce** [comment]
> 顺带一提，不知为何这样会破坏 vuepress-plugin-md-enhance 的 import 功能。

改为调用 vuepress 提供的 `renderPageContent` 后解决了，有空再研究一下是哪里不对

**Comment by DreamOfIce** [comment]
能否在 `resolvePageFileContent` 后，再调用 `extendsPageOptions`？
这样就可以免得再次渲染并写入浪费性能，并且个人认为这样更合理一些（感觉要修改 filePath 的人应该没有想修改 content 的人多）
https://github.com/vuepress/vuepress-next/blob/5d60575a9486195d45f4349324247fe2c6fcc090/packages/core/src/page/createPage.ts#L21-L31

**Comment by Mister-Hope** [comment]
我看了看源码我没看懂你要说什么

**Comment by DreamOfIce** [comment]
> 我看了看源码我没看懂你要说什么

意思就是能不能在读取完markdown文件后再调用extendsPageOptions，像这样：
```ts
 // resolve page file absolute path and relative path 
 const { filePath, filePathRelative } = resolvePageFilePath({ 
   app, 
   options, 
 }) 
  
 // read the raw file content according to the absolute file path 
 const content = await resolvePageFileContent({ filePath, options }) 

 // plugin hook: extendsPageOptions 
 await app.pluginApi.hooks.extendsPageOptions.process(options, app) 
```

**Comment by Mister-Hope** [comment]
那这个hook有什么用？你先好好理解一下这个hook的设计目标吧。

**Comment by DreamOfIce** [comment]
再读了一遍文档，抱歉是我以前没正确理解
